File: Data Structures & Algorithms/three-integer-sum/submission-15.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()

        res=[]

        for i in range(len(nums)):
            if i>0 and nums[i] ==nums[i-1]:
                continue
            l = i+1
            r = len(nums)-1
            while l < r:
                current = nums[l] + nums[r] + nums[i]
                if current == 0:
                    if [nums[l],nums[r],nums[i]] in res:
                        l += 1
                        continue
                    res.append([nums[l],nums[r],nums[i]])
                    l += 1
                    r -= 1
                if current < 0:
                    l += 1
                if current > 0:
                    r -= 1

        return res



        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-15.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        // approach: prefix and suffix sum
        // prefixes = []
        // prefix = 1

        vector<int> res(nums.size());

        vector<int> prefixes(nums.size(), 1);
        int prefix = 1;
        for (int i = 0; i < nums.size(); i++) {
            prefixes[i] = prefix;
            prefix *= nums[i];
        }

        vector<int> suffixes(nums.size(), 1);
        int suffix = 1;
        for (int i = nums.size()-1; i >-1; i--) {
            suffixes[i] = suffix;
            suffix *= nums[i];
            res[i] = suffixes[i] * prefixes[i];
        }

        return res;

    }
};